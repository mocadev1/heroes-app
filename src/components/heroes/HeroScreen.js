import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router'
import { getHeroById } from '../../selectors/getHeroById';

const heroesImages = require.context('../../assets/heroes', true);

const HeroScreen = ({ history }) => {

    const { heroId } = useParams();

    const hero = useMemo(() => getHeroById( heroId ), [ heroId ])

    if ( !hero ) {
        return <Redirect to="/" />;
    }

    const handleReturn = () => {
        if ( history.length <=2 ) {
            history.push('/');
        } else {
            history.goBack();
        }
    };

    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;
    

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img
                    // src={`../assets/heroes/${ heroId }.jpg`} // Desde public/assets
                    src={ heroesImages(`./${ heroId }.jpg`).default }
                    alt={ superhero }
                    className="img-thumbnail animate__animated animate__fadeInLeft"
                />
            </div>


            <div className="col-8 animate__animated animate__fadeIn">

                <h3> { superhero } </h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <strong> Alter ego: </strong> { alter_ego } </li>
                    <li className="list-group-item"> <strong> Publisher: </strong> { publisher } </li>
                    <li className="list-group-item"> <strong> First appearance: </strong> { first_appearance } </li>
                </ul>

                <h5> Characters </h5>
                <p> { characters } </p>

                <button
                    className="btn btn-outline-info"
                    onClick={ handleReturn }
                >
                    Return
                </button>


                
            </div>

            
        </div>
    )
}

export default HeroScreen
