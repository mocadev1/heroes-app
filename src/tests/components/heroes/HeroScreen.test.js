import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router';
import HeroScreen from '../../../components/heroes/HeroScreen';

describe('<HeroScreen /> tests', () => {
    
    const history = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn(),
    };
    
    
    test('should render the Redirect component whether there is no arguments in URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen history={ history }/>
            </MemoryRouter>
        );

        expect( wrapper.find('Redirect').exists() ).toBe(true);
    });

    test('should show a hero whether a parameter exists and is found', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-daredevil']}>
                <Route path="/hero/:heroId" component={ HeroScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').text().trim() ).toBe('Matthew Michael Murdock');
    });

    test('should go back to the previous screen with PUSH', () => {

        const history = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-daredevil']}>
                <Route
                    path="/hero/:heroId"
                    component={ (props) => <HeroScreen history={ history } />} />
            </MemoryRouter>
        );
        

        wrapper.find('button').prop('onClick')();

        expect( history.push ).toHaveBeenCalledWith('/');
        expect( history.goBack ).not.toHaveBeenCalled();

        
    });

    test('should go back to the previous screen with GOBACK', () => {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-daredevil']}>
                <Route
                    path="/hero/:heroId"
                    component={ (props) => <HeroScreen history={ history } />} />
            </MemoryRouter>
        );
        

        wrapper.find('button').prop('onClick')();

        expect( history.push ).not.toHaveBeenCalled();
        expect( history.goBack ).toHaveBeenCalled();

        
    });

    test('should call Redirect component whether does not exist', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-daredevil-fadfkaÃ±dfj']}>
                <Route
                    path="/hero/:heroId"
                    component={ (props) => <HeroScreen history={ history } />} />
            </MemoryRouter>
        );


        expect( wrapper.html() ).toBe('');
        
        
    });
    
    
    
    
    
});
