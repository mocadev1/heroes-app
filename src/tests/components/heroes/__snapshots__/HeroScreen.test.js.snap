// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeroScreen /> tests should show a hero whether a parameter exists and is found 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Daredevil"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="../assets/heroes/marvel-daredevil.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
       
      Daredevil
       
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
         
        <strong>
           Alter ego: 
        </strong>
         
        Matthew Michael Murdock
         
      </li>
      <li
        className="list-group-item"
      >
         
        <strong>
           Publisher: 
        </strong>
         
        Marvel Comics
         
      </li>
      <li
        className="list-group-item"
      >
         
        <strong>
           First appearance: 
        </strong>
         
        Daredevil #1
         
      </li>
    </ul>
    <h5>
       Characters 
    </h5>
    <p>
       
      Matthew Michael Murdock
       
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
